x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/FeatureNotEnabledException.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.security.support;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.rest.RestStatus;
+
+public class FeatureNotEnabledException extends ElasticsearchException {
+
+    public static final String DISABLED_FEATURE_METADATA = "es.disabled.feature";
+
+    /**
+     * The features names here are constants that form part of our API contract.
+     * Callers (e.g. Kibana) may be dependent on these strings. Do not change them without consideration of BWC.
+     */
+    public enum Feature {
+        TOKEN_SERVICE("security_tokens"),
+        API_KEY_SERVICE("api_keys");
+
+        private final String featureName;
+
+        Feature(String featureName) {
+            this.featureName = featureName;
+        }
+    }
```

Comment:
I wonder whether we should give `Feature` more attention to make it more formalized, because it seems to be an useful concept for other places as well. By formalization, I mean something like, have it in a separate class, place it in a shared package which is more accessible, name each feature more consistently (more on this below). 

Including Feature, we now have three different but related concepts: Feature, Setting, Plugin. Plugin seems to be the highest level, in that it provides one or multiple Features and Settings. I am not sure whether Feature has an 1-to-1 relationship with a `Setting`. If so, I feel the associated Setting should be reflected in this class and this could be useful for users to act accordingly. 

If the relationship between Feature and Setting is more complex, e.g. a feature corresponds to multiple Settings working together, it may need to be more carefully defined and promoted to be a first level concept. I am a bit surprised that Feature is not already an existing concept since both Setting and Plugin sound too technical when talking to users, while Feature seem to be a more accessible term.

I am aware that it is not something inherent to this PR so we don't need to solve it all with this PR. One possible actionable item is to move this class to a more common package for reuse by places other than security.

Reply:
Yes, no, maybe...

We do have "feature" concepts throughout the codebase. The obvious example being `isXyzzyAllowed()` on XPackLicenseState. Those are, roughly speaking, features.

I can imagine a world when we had strong model of "Feature" and the license checks where all `isFeatureAllowed(Feature feature)`, and elsewhere `isFeatureEnabled(Feature)`, but we don't and I don't think anyone wants to.

The reason I added an enum here is
1. So that the feature names were static & defined in a single place, so that clients could rely on the names being a well defined set (at least, per version).
2. I could have solved that with constant Strings, but then the exception constructor would have been `(String, String, Object ...)` and there would be ambiguity about which `String` was the feature name and which was the message.

I think there's something to your suggestion, but _progress over perfection_ - to me, this is a straight forward solution to the problem at hand, and we can evolve it into something else in the future if we need.

Reply:
Enum is good (at least for now). I agree that we don't wanna stall on something that is not in scope of the problem at hand. My actual question is whether we should move this file to a more common module, e.g. the xpack `core` module for dependency sake, since I can see it being useful for other xpack modules besides security (similar to how license related files are in `core`).

Reply:
My personal view, is that it's a false sense of "reuse".
As it stands:
1. we don't know if this would be useful elsewhere. We can imagine places, but we have no plans to change them
2. because of (1), we also don't know which enum values would make sense in the future
3. we do know that future uses of it would be constrained by not being able to change it significantly because they'd risk breaking clients.

Based on the above, I understand the temptation to put it in core, but there's no clear value, and 1 definite limitation, so it seems like something that _feels good_ for reuse purposes, but actually isn't.

Reply:
I was thinking about using it in the "[role template validation](#52636)" if we chose to issue warning instead of error out. Since we have decided it's ok to error out, I currently have no other idea where it could be re-used. So the current form is ok with me.